pkg/attestation/attestation.go

```diff
@@ -0,0 +1,49 @@
+package attestation
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"io"
+
+	"github.com/in-toto/in-toto-golang/in_toto"
+	"github.com/secure-systems-lab/go-securesystemslib/dsse"
+	"golang.org/x/xerrors"
+)
+
+type Statement struct {
+	in_toto.Statement
+	CosignPredicateData interface{} `json:"-"`
+}
+
+// Decode returns the in-toto statement from the in-toto attestation.
+func Decode(r io.Reader) (Statement, error) {
+
+	var envelope dsse.Envelope
+	err := json.NewDecoder(r).Decode(&envelope)
+	if err != nil {
+		return Statement{}, xerrors.Errorf("failed to decode as a dsse envelope: %w", err)
+	}
+	if envelope.PayloadType != in_toto.PayloadType {
+		return Statement{}, xerrors.Errorf("invalid attestation payload type: %s", envelope.PayloadType)
+	}
+
+	decoded, err := base64.StdEncoding.DecodeString(envelope.Payload)
+	if err != nil {
+		return Statement{}, xerrors.Errorf("failed to decode attestation payload: %w", err)
+	}
+
+	var st Statement
+	err = json.NewDecoder(bytes.NewReader(decoded)).Decode(&st)
+	if err != nil {
+		return Statement{}, xerrors.Errorf("failed to decode attestation payload as in-toto statement: %w", err)
+	}
+
+	// When cosign creates an SBOM attestation, it stores the predicate under a "Data" key.
+	// https://github.com/sigstore/cosign/blob/938ad43f84aa183850014c8cc6d999f4b7ec5e8d/pkg/cosign/attestation/attestation.go#L39-L43
+	if _, found := st.Predicate.(map[string]interface{})["Data"]; found {
```

Comment:
It causes panic if the type assertion fails. Can we make sure the assertion works beforehand?

```suggestion
	if cosignPredicate, ok := st.Predicate.(map[string]interface{}); ok {
		data, found := cosignPredicate["Data"]
		if !found {
			return Statement{}, xerrors.Errorf("unsupported predicate format")
		} 
		st.CosignPredicateData = data
```

Also, we can define our own struct only with needed fields so that we will not have to go back and forth for marshaling/unmarshaling. We can pass `json.RawMessage` to the SBOM unmarshaler.
https://github.com/aquasecurity/trivy/blob/87b2d211ce6fd6091f106a5067768c67791074f0/pkg/sbom/attestation/attestation.go#L28-L31

```
// StatementHeader defines the common fields for all statements
type StatementHeader struct {
	PredicateType string    `json:"predicateType"`
}

// CosignPredicate specifies the format of the Custom Predicate.
type CosignPredicate struct {
	Data      json.RawMessage
}

/*
Statement binds the attestation to a particular subject and identifies the
of the predicate. This struct represents a generic statement.
*/
type Statement struct {
	StatementHeader
	// Predicate contains type speficic metadata.
	Predicate CosignPredicate `json:"predicate"`
}
```

Reply:
Great idea! It will simplify the code!